Add pre/post hooks to deploy/cleanup script and focus/skip to e2e script

mkimuram opened the pull request:
This PR adds below features to shipyard e2e scripts:

- Pre/post hooks in deploy/cleanup script
- Focus/skip in e2e script

Fixes: #508

mkimuram committed:
Allow e2e script to pass -ginkgo.skip

- Add --skip to pass through the regex to -ginkgo.skip of e2e test,
- Change --focus to pass through the regex to -ginkgo.focus as is,
  not surrounded with "\[" and "\]" for symmetry,
- Make each regex pass only when regex is not empty, to allow
  specifying not to focus nor skip.

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram committed:
Add nob to pass focus and skip to Makefile

Added `FOCUS` environment variable and `SKIP` environment variable to Makefile.

- The value of `FOCUS` environment variable is added with --focus option to
  `E2E_ARGS` and used from e2e.sh.
- The value of `SKIP` environment variable is added with --skip option to
  `E2E_ARGS` and used from e2e.sh.

ex)

```
export FOCUS="\[example\]"
export SKIP="Should be able to list"
make e2e
```

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram committed:
Add nob to pass focus and skip to github actions

Added `focus` parameter and `skip` parameter to gh-actions/e2e/action.yaml.

- The value of `focus` parameter is defined as `FOCUS` environment variable
  and used from `make e2e` command.
- The value of `skip` parameter is defined as `SKIP` environment variable
  and used from `make e2e` command.

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram committed:
Add pre/post hook for deploy and cleanup

Below 4 hook functions can be defined in the plugin file.
If the file is specified with the --plugin option of deploy.sh and cleanup.sh,
each function is called at the below event.

- deploy.sh
  - pre_deploy: called before deploy,
  - post_deploy: called after deploy.

- cleanup.sh
  - pre_cleanup: called before cleanup,
  - post_cleanup: called after cleanup.

A sample hook is placed in scripts/shared/resources/plugins/sample_hook.

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram committed:
Add nob to pass pluin to Makefile

Added `PLUGIN` environment variable to Makefile.
The value is added with --plugin option to `DEPLOY_ARGS` and `CLEANUP_ARGS`
and used from cleanup.sh, deploy.sh, and e2e.sh.

ex)

```
base_dir="/go/src/github.com/submariner-io/shipyard/"
hook_dir="scripts/shared/resources/plugins/sample_hook"
export PLUGIN="${base_dir}/${hook_dir}"
make deploy
make cleanup
```

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram committed:
Add nob to pass pluin to github actions

Added `plugin` parameter to gh-actions/e2e/action.yaml.
The value is defined as `PLUGIN` environment variable and used from
`make e2e` command.

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram commented:
As described in commmit messages, these features can be used:

- Pre/post hooks in deploy/cleanup script:
```
export PLUGIN=/go/src/github.com/submariner-io/shipyard/scripts/shared/resources/plugins/sample_hook
make deploy
make cleanup
```

- Focus/skip in e2e script:
```
export FOCUS="\[example\]"
export SKIP="Should be able to list"
make e2e
```

mkimuram commented:
@dfarrell07 

I've implemented the scripts. Please review it.

dfarrell07 commented on gh-actions/e2e/action.yaml:
```diff
@@ -12,6 +12,18 @@ inputs:
     required: false
     default: '.'
+  focus:
+    description: '-ginkgo.focus argument to passed to e2e tests'
```
```suggestion
    description: '-ginkgo.focus argument to passed to E2E tests'
```

Nit: I think we capitalize it consistently in this context, so let's continue doing so.

dfarrell07 commented on gh-actions/e2e/action.yaml:
```diff
@@ -12,6 +12,18 @@ inputs:
+    required: false
+    default: '\[.*\]'
+  skip:
+    description: '-ginkgo.skip argument to passed to e2e tests'
```
```suggestion
    description: '-ginkgo.skip argument to passed to E2E tests'
```

dfarrell07 reviewed (commented):
This is great @mkimuram, thank you! One very minor nit inline.

dfarrell07 commented:
There are also some commit message linting issues:

```
gitlint --commits origin/devel..HEAD
Commit 9d8033a5fe:
10: B1 Line exceeds max length (100>80): "export PLUGIN=/go/src/github.com/submariner-io/shipyard/scripts/shared/resources/plugins/sample_hook"

Commit aad86c5d2d:
5: B1 Line exceeds max length (86>80): "- The value of `FOCUS` environment variable is added with --focus option to `E2E_ARGS`"
7: B1 Line exceeds max length (84>80): "- The value of `SKIP` environment variable is added with --skip option to `E2E_ARGS`"
```

dfarrell07 commented:
And also some shellcheck linting issues:

```
shellcheck -e SC2154 scripts/shared/lib/*

In scripts/shared/lib/utils line 138:
    type -t $func | grep -q function || return 0
            ^---^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
    type -t "$func" | grep -q function || return 0


In scripts/shared/lib/utils line 139:
    eval $@
         ^-- SC2068: Double quote array expansions to avoid re-splitting elements.

For more information:
  https://www.shellcheck.net/wiki/SC2068 -- Double quote array expansions to ...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

mkimuram commented:
@dfarrell07 

Thank you for your review. Fixed lint and commit issues and changed e2e to E2E.
It seems that the upgrade test is failing. I'm investigating it.

dfarrell07 commented:
> It seems that the upgrade test is failing. I'm investigating it.

@mkimuram It's not related to your PR, you can ignore it.

dfarrell07 reviewed (approved):
Thanks for the quality PR @mkimuram! LGTM :+1:

tpantelis commented on gh-actions/e2e/action.yaml:
```diff
@@ -12,6 +12,18 @@ inputs:
     required: false
     default: '.'
+  focus:
+    description: '-ginkgo.focus argument to passed to E2E tests'
```
```suggestion
    description: '-ginkgo.focus argument to pass to E2E tests'
```

tpantelis commented on gh-actions/e2e/action.yaml:
```diff
@@ -12,6 +12,18 @@ inputs:
+    required: false
+    default: '\[.*\]'
+  skip:
+    description: '-ginkgo.skip argument to passed to E2E tests'
```
```suggestion
    description: '-ginkgo.skip argument to pass to E2E tests'
```

tpantelis reviewed (changes requested):
Overall looks good.

tpantelis committed:
Merge branch 'devel' into plugable-e2e
